File: packages/kits/AriaMcpKit/Tests/AriaMCPTests/FrozenPostureTests.swift
// FrozenPostureTests.swift — the frozen serve posture at the dispatcher.
//
// Mirrors packages/kits/AriaMcpKit/rust/tests/frozen_posture_tests.rs.
//
// Coverage:
//   1. EstatePosture resolution: flag wins, MOOTX01_FROZEN=1 enables, else live.
//   2. ToolMutationInventory names only reachable tools, covers the writers,
//      and classifies every reachable tool (under every opt-in flag
//      combination) into exactly one of the read set, the refused set, or
//      the command-classified set.
//   3. A frozen dispatcher refuses moot_file_memory / moot_update_memory with
//      the exact isError text and no side effect, allows moot_memory_search and
//      moot_estate_status, and reports `frozen: true`.
//   4. A frozen search then dereference writes no recall-trace rows and leaves
//      the reward mark untouched (probed through kit.markRecallUsed).
//   5. `memory` is view-only when frozen: every other command is refused
//      before the adapter runs and before session state records the call,
//      with the estate byte-identical on disk; the same delete lands live.
//
// SQLite-backed where trace rows are involved: the recall_trace table only
// exists on the SQLite backend.

import Testing
import Foundation
import GeniusLocusKit
import LocusKit
import PersistenceKit
import PersistenceKitInMemory
import PersistenceKitSQLite
@testable import AriaMCP

@Suite("EstatePosture resolution")
struct EstatePostureResolutionTests {

    @Test func absentEnvironmentIsLive() {
        #expect(EstatePosture.resolve(frozenFlag: false, environment: [:]) == .live)
    }

    @Test func environmentOneFreezes() {
        #expect(EstatePosture.resolve(frozenFlag: false, environment: ["MOOTX01_FROZEN": "1"]) == .frozen)
    }

    @Test func environmentOtherValuesStayLive() {
        for value in ["0", "true", "yes", ""] {
            #expect(EstatePosture.resolve(frozenFlag: false, environment: ["MOOTX01_FROZEN": value]) == .live,
                    "MOOTX01_FROZEN=\(value) must not freeze; only \"1\" does")
        }
    }

    @Test func flagWinsOverEnvironment() {
        #expect(EstatePosture.resolve(frozenFlag: true, environment: ["MOOTX01_FROZEN": "0"]) == .frozen)
        #expect(EstatePosture.resolve(frozenFlag: true, environment: [:]) == .frozen)
    }

    @Test func logLineAndRefusalTextArePinned() {
        // Both strings are contract text shared byte-for-byte with the Rust port.
        #expect(EstatePosture.frozenLogLine
                == "FROZEN: no background workers, no recall traces, mutating tools refused")
        #expect(EstatePosture.refusalMessage(tool: "moot_file_memory")
                == "estate is frozen (serve --frozen): moot_file_memory is a mutating tool and was refused")
        #expect(EstatePosture.frozen.statusValue == "true")
        #expect(EstatePosture.live.statusValue == "false")
    }
}

@Suite("ToolMutationInventory")
struct ToolMutationInventoryTests {

    /// Every tool name a serve launched with `environment` can dispatch: the
    /// advertised projection.
    private func reachable(_ environment: [String: String]) -> Set<String> {
        Set(ToolProjection.tools(environment: environment).map(\.name))
    }

    /// Every flag combination a serve can be launched with.
    private static let flagCombinations: [[String: String]] = {
        var combinations: [[String: String]] = []
        for vault in ["1", "0"] {
            for memory in ["0", "1"] {
                combinations.append(["MOOTX01_VAULT": vault, "MOOTX01_MEMORY_TOOL": memory])
            }
        }
        return combinations
    }()

    /// Every opt-in on: the widest surface a serve can dispatch.
    private static let widest = ["MOOTX01_VAULT": "1", "MOOTX01_MEMORY_TOOL": "1"]

    /// Every name in the inventory must be a tool a serve can really
    /// dispatch; a renamed or retired tool must fail here, not silently stop
    /// being refused (or stop being allowed).
    @Test func inventoryNamesOnlyReachableTools() {
        let real = reachable(Self.widest)
        let classified = ToolMutationInventory.frozenReadTools
            .union(ToolMutationInventory.frozenRefusedTools)
            .union(ToolMutationInventory.commandClassifiedTools)
        let stale = classified.subtracting(real)
        #expect(stale.isEmpty, "inventory names tool(s) no serve can dispatch: \(stale.sorted())")
    }

    /// The structural guarantee: a tool a frozen serve can dispatch is in
    /// exactly one of the read set, the refused set, or the command-classified
    /// set, under every opt-in flag combination. A new tool in none of them
    /// fails here with its name.
    @Test func everyReachableToolIsInExactlyOneFrozenSet() {
        for environment in Self.flagCombinations {
            for name in reachable(environment).sorted() {
                let buckets = [
                    ToolMutationInventory.frozenReadTools.contains(name),
                    ToolMutationInventory.frozenRefusedTools.contains(name),
                    ToolMutationInventory.commandClassifiedTools.contains(name),
                ].filter { $0 }.count
                #expect(buckets == 1,
                        "\(name) is in \(buckets) frozen sets under \(environment); every reachable tool must be in exactly one of frozenReadTools / frozenRefusedTools / commandClassifiedTools")
            }
        }
    }

    @Test func writersAreRefusedAndReadersAreNot() {
        let refused = ToolMutationInventory.frozenRefusedTools
        for tool in ["moot_file_memory", "moot_update_memory", "moot_erase_memory",
                     "moot_dream", "moot_json_import", "moot_file_fact"] {
            #expect(refused.contains(tool), "\(tool) must be refused when frozen")
        }
        for tool in ["moot_memory_search", "moot_estate_status", "moot_memory_get", "moot_recall_precise",
                     "moot_lens_concepts", "moot_estate_ping", "moot_drain_status", "moot_dataset_query"] {
            #expect(!refused.contains(tool), "\(tool) is a read and must stay callable when frozen")
            #expect(ToolMutationInventory.frozenReadTools.contains(tool),
                    "\(tool) is a read and must be in the explicit read set")
        }
        // The three sets are disjoint: a tool has exactly one tier.
        #expect(ToolMutationInventory.additiveWriteTools.isDisjoint(with: ToolMutationInventory.mutationTools))
        #expect(ToolMutationInventory.mutationTools.isDisjoint(with: ToolMutationInventory.destructiveTools))
        #expect(ToolMutationInventory.additiveWriteTools.isDisjoint(with: ToolMutationInventory.destructiveTools))
    }

    @Test func memoryIsCommandClassifiedWithViewAsItsOnlyRead() {
        #expect(ToolMutationInventory.frozenReadCommands["memory"] == ["view"])
        #expect(ToolMutationInventory.commandClassifiedTools == ["memory"])
        #expect(!ToolMutationInventory.frozenRefusedTools.contains("memory"),
                "memory is classified by command, never by name")
        #expect(!ToolMutationInventory.frozenReadTools.contains("memory"))
    }
}

/// `.serialized`: filesystem interaction; keep sequential to avoid temp-file
/// collisions.
@Suite("Frozen dispatcher", .serialized)
struct FrozenDispatcherTests {

    // MARK: - Helpers

    private func tempDBURL() throws -> URL {
        let dir = FileManager.default.temporaryDirectory
            .appendingPathComponent("FrozenPostureTests-\(UUID().uuidString)", isDirectory: true)
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir.appendingPathComponent("test.sqlite")
    }

    /// Open a SQLite-backed GLK estate and return (kit, handle). Dispatchers
    /// are built per test so live and frozen can share one estate.
    private func openSQLiteEstate(url: URL) async throws -> (GeniusLocusKit, EstateHandle) {
        let kit = GeniusLocusKit()
        let owner = OwnerCredentials(ownerIdentifier: "frozen-posture-tests")
        let configuration = EstateConfiguration(
            estateID: UUID(),
            backend: .sqlite(url: url, busyTimeout: 5.0)
        )
        let storage = try SQLiteStorage(configuration: configuration)
        _ = try await LocusKit.Estate.create(storage: storage, owner: owner)
        let handle = try await kit.open(storage: storage, owner: owner,
                                        identityKeyStore: InMemoryEstateIdentityKeyStore())
        return (kit, handle)
    }

    private func firstText(_ result: JSONValue) -> String {
        result.objectValue?["content"]?.arrayValue?.first?
            .objectValue?["text"]?.stringValue ?? ""
    }

    private func isError(_ result: JSONValue) -> Bool {
        result.objectValue?["isError"]?.boolValue ?? false
    }

    /// The on-disk estate: the main database file plus its WAL sibling (the
    /// kit runs SQLite in WAL mode, so a write that has not been checkpointed
    /// lives in `-wal`). Two snapshots that compare equal prove no byte of
    /// committed or pending state changed between them.
    private func estateBytes(_ url: URL) throws -> [Data] {
        let wal = URL(fileURLWithPath: url.path + "-wal")
        let walBytes = FileManager.default.fileExists(atPath: wal.path) ? try Data(contentsOf: wal) : Data()
        return [try Data(contentsOf: url), walBytes]
    }

    private func fileMemory(_ dispatcher: ToolDispatcher, content: String, location: String) async throws -> String {
        let result = try await dispatcher.dispatch(
            name: "moot_file_memory",
            arguments: .object([
                "content": .string(content),
                "subject": .string(String(content.prefix(120))),
                "location": .string(location),
            ])
        )
        let idLine = firstText(result).split(separator: "\n").first.map(String.init) ?? ""
        let id = idLine.replacingOccurrences(of: "filed memory ", with: "")
        #expect(!id.isEmpty, "filed memory id must be non-empty; got: \(firstText(result))")
        return id
    }

    private func search(_ dispatcher: ToolDispatcher, query: String) async throws -> JSONValue {
        try await dispatcher.dispatch(name: "moot_memory_search",
                                      arguments: .object(["query": .string(query)]))
    }

    private func status(_ dispatcher: ToolDispatcher) async throws -> String {
        firstText(try await dispatcher.dispatch(name: "moot_estate_status", arguments: .object([:])))
    }

    // MARK: - Construction

    @Test func postureComesFromFlagThenEnvironment() async throws {
        let url = try tempDBURL()
        defer { try? FileManager.default.removeItem(at: url.deletingLastPathComponent()) }
        let (kit, handle) = try await openSQLiteEstate(url: url)

        #expect(ToolDispatcher(kit: kit, handle: handle, environment: [:]).posture == .live)
        #expect(ToolDispatcher(kit: kit, handle: handle,
                               environment: ["MOOTX01_FROZEN": "1"]).posture == .frozen)
        // Explicit posture (the --frozen flag) wins over the environment twin.
        let explicit = ToolDispatcher(kit: kit, handle: handle,
                                      environment: ["MOOTX01_FROZEN": "0"], posture: .frozen)
        #expect(explicit.posture == .frozen)
        // Derived dispatchers carry the posture unchanged.
        #expect(explicit.registering(handle).posture == .frozen)
        #expect(explicit.withMonitoringControl(nil).posture == .frozen)
    }


    // MARK: - Command-classified tool: memory is view-only when frozen

    /// `memory` is classified per call: `view` proceeds and reads; every
    /// other command, and a missing or unknown one, is refused before the
    /// adapter runs and before session state records the call, with the
    /// estate byte-identical on disk. The adapter itself is posture-blind:
    /// the same `delete` lands through a live dispatcher.
    @Test func frozenMemoryToolIsViewOnly() async throws {
        let url = try tempDBURL()
        defer { try? FileManager.default.removeItem(at: url.deletingLastPathComponent()) }
        let (kit, handle) = try await openSQLiteEstate(url: url)
        let memoryOn = ["MOOTX01_MEMORY_TOOL": "1"]
        let live = ToolDispatcher(kit: kit, handle: handle, environment: memoryOn)
        let frozen = ToolDispatcher(kit: kit, handle: handle, environment: memoryOn, posture: .frozen)
        let path = "/memories/frozen-notes.txt"

        // One file created live, so view has something to read and delete a target.
        let created = try await live.dispatch(name: "memory", arguments: .object([
            "command": .string("create"), "path": .string(path),
            "file_text": .string("frozen posture view-only test"),
        ]))
        #expect(firstText(created).contains("File created successfully"), "precondition; got: \(firstText(created))")

        let before = try estateBytes(url)
        let callsBefore = await frozen.modeSessionState.totalCallCount
        let mutating: [(command: String?, arguments: [String: JSONValue])] = [
            ("create", ["path": .string("/memories/other.txt"), "file_text": .string("must not land")]),
            ("str_replace", ["path": .string(path), "old_str": .string("view-only"), "new_str": .string("must not land")]),
            ("insert", ["path": .string(path), "insert_line": .integer(0), "insert_text": .string("must not land")]),
            ("delete", ["path": .string(path)]),
            ("rename", ["old_path": .string(path), "new_path": .string("/memories/renamed.txt")]),
            ("frobnicate", ["path": .string(path)]),
            (nil, ["path": .string(path)]),
        ]
        for (command, arguments) in mutating {
            var args = arguments
            if let command { args["command"] = .string(command) }
            let result = try await frozen.dispatch(name: "memory", arguments: .object(args))
            #expect(isError(result), "memory \(command ?? "(missing)") must be refused when frozen; got: \(firstText(result))")
            #expect(firstText(result) == EstatePosture.refusalMessage(tool: "memory", command: command))
        }
        #expect(try estateBytes(url) == before, "refused memory commands must leave the estate byte-identical on disk")
        let callsAfterRefusals = await frozen.modeSessionState.totalCallCount
        #expect(callsAfterRefusals == callsBefore, "a refused memory command must not be recorded in session state")

        // view proceeds and reads the live-created file; the dispatcher records it.
        let view = try await frozen.dispatch(name: "memory",
                                             arguments: .object(["command": .string("view"), "path": .string(path)]))
        #expect(!isError(view) && firstText(view).contains("frozen posture view-only test"),
                "memory view is a read and must work when frozen; got: \(firstText(view))")
        let callsAfterView = await frozen.modeSessionState.totalCallCount
        #expect(callsAfterView == callsBefore + 1, "a view the dispatcher lets through is recorded")

        // The adapter is posture-blind: the same delete lands live.
        let deleted = try await live.dispatch(name: "memory",
                                              arguments: .object(["command": .string("delete"), "path": .string(path)]))
        #expect(firstText(deleted).hasPrefix("Successfully deleted"), "live delete must still work; got: \(firstText(deleted))")
        let gone = try await frozen.dispatch(name: "memory",
                                             arguments: .object(["command": .string("view"), "path": .string(path)]))
        #expect(firstText(gone).contains("does not exist"), "the deleted file must be gone; got: \(firstText(gone))")
    }

    // MARK: - Read path writes nothing — BLOCKED (v2 dropped dereference reward-marking)
    //
    // v1: live search seeds trace rows; a frozen search adds none, and a
    // frozen dereference (`moot_memory_get`) leaves the reward mark
    // untouched — probed via `kit.markRecallUsed` returning a non-zero
    // "still eligible to be flipped" count. This discriminates frozen from
    // live: a LIVE dereference is expected to mark the row (make
    // `markRecallUsed` return 0 the second time), while a FROZEN one must
    // not.
    //
    // In v2, `moot_memory_get`'s production path
    // (`AriaV2MemoryOperations.get`, AriaV2MemoryOperations.swift:756-782)
    // calls `context.usageLedger.recordDereferenced(...)` at line 767, but
    // every concrete `AriaV2MemoryUsageLedger` wired to it is a no-op:
    // the default implementation is an empty body
    // (`public func recordDereferenced(...) async {}`,
    // AriaV2MemoryOperations.swift:50), and the dispatcher's own
    // `DispatcherV2MemoryUsageLedger.recordDereferenced`
    // (ToolDispatch.swift:37) explicitly discards its arguments
    // (`_ = (memoryIDs, estateID, callerID, at)`) with a doc comment stating
    // "Reward marking remains owned by the established typed GLK path; this
    // adapter does not invent a second session store or mutate recall state
    // during a read." The only code that actually calls `kit.markRecallUsed`
    // on a dereference is `noteUsage(_:handle:)` (ToolDispatch.swift:2671-2688),
    // called exclusively from the dead legacy `runMemoryGet`
    // (ToolDispatch.swift:2402, line 2611), unreachable from
    // `ToolDispatcher.dispatch(name:arguments:)`.
    //
    // The result: `moot_memory_get` never marks the reward bit in v2,
    // whether the dispatcher is live or frozen. The v1 probe
    // (`unmarked > 0` after a dereference) would now be true in BOTH
    // postures, so it no longer discriminates frozen behavior from live
    // behavior — the property this case exists to prove (frozen is MORE
    // restrictive than live here) cannot be demonstrated against v2, because
    // live is no longer less restrictive on this path. Do not delete; do
    // not weaken to pass.

    @Test(.disabled("BLOCKED ON ID SPELLING, not on the reward loop. The loop is restored: DispatcherV2MemoryUsageLedger.recordDereferenced now calls kit.markRecallUsed under the four v1 conditions (live posture only, surfaced in this session, both storage spellings, a fresh wall clock later than the dispatch instant), and the four mutation verbs call it too, so v1's seven sites are covered. This case still cannot run because its own probe at :361 calls markRecallUsed(target: id) with the canonical-lowercase id taken from the v2 response text, while trace rows are keyed by the stored drawer spelling and GeniusLocusKit.markRecallUsed (VerbSurface.swift:1773-1785) passes target through verbatim with no normalisation — the same casing family recorded in the node-motion block reasons. The restoration code tries both spellings; the probe tries one. Unblocks with the public-id versus storage-id reconciliation. Do not delete; do not weaken to pass."))
    func frozenSearchThenDereferenceLeavesTracesUnmarked() async throws {
        let url = try tempDBURL()
        defer { try? FileManager.default.removeItem(at: url.deletingLastPathComponent()) }
        let (kit, handle) = try await openSQLiteEstate(url: url)
        let live = ToolDispatcher(kit: kit, handle: handle, environment: [:])
        let frozen = ToolDispatcher(kit: kit, handle: handle, environment: [:], posture: .frozen)

        let id = try await fileMemory(live, content: "frozen trace reward test", location: "trace-room")
        _ = try await search(live, query: "frozen trace reward")
        let seeded = try await kit.countRecallTraces(handle)
        #expect(seeded > 0, "live search must seed trace rows")

        // Frozen search: surfaces the drawer, writes no trace row.
        let frozenSearch = try await search(frozen, query: "frozen trace reward")
        #expect(firstText(frozenSearch).contains(id))
        #expect(try await kit.countRecallTraces(handle) == seeded,
                "a frozen search must not write recall-trace rows")

        // Frozen dereference: succeeds, marks nothing.
        let get = try await frozen.dispatch(name: "moot_memory_get", arguments: .object(["memory_id": .string(id)]))
        #expect(!isError(get), "moot_memory_get is a read and must work when frozen; got: \(get)")
        let unmarked = try await kit.markRecallUsed(handle, target: id, now: Date())
        #expect(unmarked > 0,
                "the seeded rows must still be unmarked after a frozen dereference (probe flipped \(unmarked))")
        #expect(try await kit.countRecallTraces(handle) == seeded)
    }

    // MARK: - moot_synthesize is a read under frozen

    /// `moot_synthesize` reads candidates and generates text; it writes no
    /// drawer, packet, journal, meta, trace, or reward. A frozen dispatcher
    /// must let it through, and the estate must be byte-identical before and
    /// after the call. FRZ-3: moved from mutationTools to frozenReadTools.
    @Test func frozenSynthesizeProceedsAndEstateIsUnchanged() async throws {
        let url = try tempDBURL()
        defer { try? FileManager.default.removeItem(at: url.deletingLastPathComponent()) }
        let (kit, handle) = try await openSQLiteEstate(url: url)
        let live = ToolDispatcher(kit: kit, handle: handle, environment: [:])
        let frozen = ToolDispatcher(kit: kit, handle: handle, environment: [:], posture: .frozen)

        _ = try await fileMemory(live, content: "carbon compounds synthesis test", location: "synth-room")
        let before = try estateBytes(url)

        let result = try await frozen.dispatch(
            name: "moot_synthesize",
            arguments: .object(["query": .string("carbon compounds"), "limit": .integer(5)]))
        #expect(!isError(result),
                "moot_synthesize is a read and must work when frozen; got: \(firstText(result))")
        #expect(try estateBytes(url) == before,
                "moot_synthesize must leave the estate byte-identical on disk")
        // The frozen refused-set must no longer name moot_synthesize.
        #expect(!ToolMutationInventory.frozenRefusedTools.contains("moot_synthesize"),
                "moot_synthesize must not be in the refused set after FRZ-3")
        #expect(ToolMutationInventory.frozenReadTools.contains("moot_synthesize"),
                "moot_synthesize must be in the explicit read set after FRZ-3")
    }

}

File: packages/kits/AriaMcpKit/rust/tests/frozen_posture_tests.rs
//! Frozen posture tests — Rust twin of Swift `FrozenPostureTests.swift`.
//!
//! Coverage:
//!   1. A frozen dispatcher refuses `moot_file_memory` / `moot_update_memory`
//!      with the v2 estate_frozen error code and no side effect, allows
//!      `moot_memory_search` and `moot_estate_status`, and leaves the memory
//!      count unchanged; a live dispatcher accepts writes.
//!   2. A frozen search then dereference writes no recall-trace rows and
//!      leaves the reward mark untouched (probed through `mark_recall_used`,
//!      which returns the number of rows it flips).
//!   3. `moot_synthesize` reads candidates and generates text; it writes no
//!      drawer, packet, journal, meta, trace, or reward. A frozen dispatcher
//!      must let it through, and the estate must be byte-identical before and
//!      after the call. FRZ-3: moved from MUTATION_TOOLS to FROZEN_READ_TOOLS.
//!
//! The posture is injected with `with_posture`, never through the process
//! environment: std::env is process-global and the test runner is parallel.
//! `EstatePosture` resolution itself is covered by the unit tests in
//! `estate_posture.rs`.

use aria_mcp::dispatcher::Dispatcher;
use aria_mcp::estate_posture::EstatePosture;
use aria_mcp::estate_registry::EstateRegistry;
use aria_mcp::jsonrpc::JSONRPCRequest;
use genius_locus_kit::recall::{
    GLKRecallMode, GLKRecallRequest, GLKRecallScoring, RecallFallbackPolicy, RecallOrigin,
};
use locus_kit::drawer_operational::CaptureChannel;
use locus_kit::estate_types::LatticeAnchor;
use locus_kit::filter::{Filter, RecallFrame};
use locus_kit::frames::CaptureFrame;
use serde_json::Value as JsonValue;
use uuid::Uuid;

// Epoch MILLISECONDS for capture/recall `now`; ISO twins for the reward window.
const NOW: i64 = 1_700_000_000_000;
const NOW_ISO: &str = "2023-11-14T22:13:20Z";
const SINCE_FLOOR: &str = "0000-01-01T00:00:00Z";

/// Per-estate temp directory so the queue sibling is unique (mirrors
/// persistence_tests). The file does not exist yet; `new_sqlite` creates it.
fn temp_sqlite_path(label: &str) -> String {
    let dir = std::env::temp_dir().join(format!("aria_mcp_frozen_{}_{}", label, Uuid::new_v4()));
    std::fs::create_dir_all(&dir).expect("create per-estate temp dir");
    dir.join("estate.sqlite").to_string_lossy().into_owned()
}

fn make_dispatcher(registry: EstateRegistry, posture: EstatePosture) -> Dispatcher {
    Dispatcher::new(registry, "ARIA_MCP_Rust", "test", "test-serial", None).with_posture(posture)
}

fn tools_call(dispatcher: &Dispatcher, tool: &str, args: JsonValue) -> JsonValue {
    let raw = serde_json::json!({
        "jsonrpc": "2.0", "id": 1, "method": "tools/call",
        "params": { "name": tool, "arguments": args }
    });
    let request = JSONRPCRequest::decode(&raw).expect("request must decode");
    let response = dispatcher.handle(&request);
    serde_json::to_value(&response).expect("response must serialize")
}

fn is_error(response: &JsonValue) -> bool {
    response["result"]["isError"].as_bool().unwrap_or(false)
}

/// Capture a drawer straight through the coordinator (the estate is shared
/// with the dispatcher built from this registry) and return its id.
fn capture(registry: &EstateRegistry, content: &str) -> String {
    let coord = registry.coord.lock().unwrap();
    let frame = CaptureFrame::new(
        content,
        CaptureChannel::Typed,
        "frozen-room",
        LatticeAnchor::udc("0"),
        "frozen-posture-tests",
        "test-embed-v1",
    );
    coord.capture(&registry.default.handle, frame, NOW).expect("capture").id
}

/// One EXTERNAL recall so the surfaced drawer gets recall-trace rows — the
/// seeded state a frozen search must leave untouched.
fn recall_writing_traces(registry: &EstateRegistry) {
    let coord = registry.coord.lock().unwrap();
    let req = GLKRecallRequest::new(
        RecallFrame::new(vec![Filter::Unconfirmed]),
        GLKRecallMode::UnionBest,
        GLKRecallScoring::MatrixAware,
        50,
        RecallFallbackPolicy::FailClosed,
        RecallOrigin::External,
    )
    .with_trace_limit(10);
    coord.recall_scored(&registry.default.handle, req, NOW).expect("recall writing traces");
}

fn count_traces(registry: &EstateRegistry) -> usize {
    let coord = registry.coord.lock().unwrap();
    coord.count_recall_traces(&registry.default.handle).expect("count_recall_traces")
}

/// Rows flipped by the reward mark: non-zero proves the rows were still
/// unmarked when this probe ran.
fn mark_used(registry: &EstateRegistry, target: &str) -> usize {
    let coord = registry.coord.lock().unwrap();
    coord
        .mark_recall_used(&registry.default.handle, target, SINCE_FLOOR, NOW_ISO)
        .expect("mark_recall_used")
}

#[test]
fn frozen_refuses_writers_and_allows_reads() {
    let path = temp_sqlite_path("refusal");
    // Seed one drawer through a registry that is then moved into the frozen
    // dispatcher, so update has a real target and search has a hit.
    let registry = EstateRegistry::new_sqlite(&path, "frozen-tests").expect("open");
    let id = capture(&registry, "frozen posture refusal test");
    let frozen = make_dispatcher(registry, EstatePosture::Frozen);

    // Estate status is a read tool: must succeed even when frozen.
    let before_status = tools_call(&frozen, "moot_estate_status", serde_json::json!({}));
    assert!(!is_error(&before_status), "estate_status is a read and must work when frozen; got: {before_status}");
    let before_count = before_status["result"]["structuredContent"]["data"]["memory_count"]
        .as_u64()
        .expect("memory_count must be a u64");

    // moot_file_memory refused: v2 returns estate_frozen error code in structuredContent.
    let file = tools_call(&frozen, "moot_file_memory",
        serde_json::json!({"content": "must not land", "subject": "must not land", "location": "frozen-room"}));
    assert!(is_error(&file), "moot_file_memory must be refused when frozen; got: {file}");
    assert_eq!(
        file["result"]["structuredContent"]["error"]["code"], "estate_frozen",
        "refusal code must be estate_frozen; got: {file}"
    );

    // moot_update_memory refused the same way (v2 args: memory_id + set_subject mutation).
    let update = tools_call(&frozen, "moot_update_memory",
        serde_json::json!({"memory_id": id, "mutation": "set_subject", "subject": "must not land"}));
    assert!(is_error(&update), "moot_update_memory must be refused when frozen; got: {update}");
    assert_eq!(
        update["result"]["structuredContent"]["error"]["code"], "estate_frozen",
        "refusal code must be estate_frozen; got: {update}"
    );

    // No partial side effect: memory count must be unchanged after refusals.
    let after_status = tools_call(&frozen, "moot_estate_status", serde_json::json!({}));
    assert!(!is_error(&after_status));
    let after_count = after_status["result"]["structuredContent"]["data"]["memory_count"]
        .as_u64()
        .expect("memory_count must be a u64");
    assert_eq!(after_count, before_count, "refusal must not file anything; counts diverged");

    // Reads keep working: search surfaces the seeded drawer via its memory_id.
    let search = tools_call(&frozen, "moot_memory_search", serde_json::json!({"query": "frozen posture refusal"}));
    assert!(!is_error(&search), "moot_memory_search is a read and must work when frozen; got: {search}");
    let results = search["result"]["structuredContent"]["data"]["results"]
        .as_array()
        .expect("results must be an array");
    assert!(
        results.iter().any(|r| r["fetch"]["arguments"]["memory_id"].as_str() == Some(id.as_str())),
        "frozen search must surface the seeded drawer; got: {search}"
    );

    let _ = std::fs::remove_file(&path);
}

#[test]
fn live_dispatcher_reports_no_frozen_refusals() {
    let live = make_dispatcher(EstateRegistry::new_inmemory(), EstatePosture::Live);
    // Estate status returns a valid structured response for a live dispatcher.
    let status = tools_call(&live, "moot_estate_status", serde_json::json!({}));
    assert!(!is_error(&status), "estate_status must succeed for a live dispatcher; got: {status}");
    // A live dispatcher must accept write tools without issuing estate_frozen refusals.
    let filed = tools_call(&live, "moot_file_memory",
        serde_json::json!({"content": "lands live", "subject": "lands live", "location": "live-room"}));
    assert!(!is_error(&filed), "a live dispatcher must file without refusal; got: {filed}");
}

#[test]
fn frozen_search_then_dereference_leaves_traces_unmarked() {
    let path = temp_sqlite_path("traces");
    let seed = EstateRegistry::new_sqlite(&path, "frozen-tests").expect("open");
    let id = capture(&seed, "frozen trace reward test");
    recall_writing_traces(&seed);
    let seeded = count_traces(&seed);
    assert!(seeded > 0, "external recall must seed trace rows");
    // The probe registry reopens the same file: trace counting and the
    // reward mark are SQL over the shared database.
    let probe = EstateRegistry::new_sqlite(&path, "frozen-tests").expect("reopen");
    let frozen = make_dispatcher(seed, EstatePosture::Frozen);

    // Frozen search: surfaces the drawer in v2 structured results, writes no trace row.
    let search = tools_call(&frozen, "moot_memory_search", serde_json::json!({"query": "frozen trace reward"}));
    assert!(!is_error(&search), "frozen search must succeed; got: {search}");
    let results = search["result"]["structuredContent"]["data"]["results"]
        .as_array()
        .expect("results must be an array");
    assert!(
        results.iter().any(|r| r["fetch"]["arguments"]["memory_id"].as_str() == Some(id.as_str())),
        "frozen search must surface the seeded drawer; got: {search}"
    );
    assert_eq!(count_traces(&probe), seeded, "a frozen search must not write recall-trace rows");

    // Frozen dereference: succeeds using v2 memory_id arg, marks nothing.
    let get = tools_call(&frozen, "moot_memory_get", serde_json::json!({"memory_id": id}));
    assert!(!is_error(&get), "moot_memory_get is a read and must work when frozen; got: {get}");
    let unmarked = mark_used(&probe, &id);
    assert!(unmarked > 0, "the seeded rows must still be unmarked after a frozen dereference (probe flipped {unmarked})");
    assert_eq!(count_traces(&probe), seeded);

    let _ = std::fs::remove_file(&path);
}

/// `moot_synthesize` reads candidates and generates text; it writes no drawer,
/// packet, journal, meta, trace, or reward. A frozen dispatcher must let it
/// through, and the estate must be byte-identical before and after the call.
/// FRZ-3: moved from MUTATION_TOOLS to FROZEN_READ_TOOLS.
#[test]
fn frozen_synthesize_proceeds_and_estate_is_unchanged() {
    use aria_mcp::tool_mutation_inventory::{FROZEN_READ_TOOLS, is_frozen_refused};
    let path = temp_sqlite_path("synthesize");
    let registry = EstateRegistry::new_sqlite(&path, "frozen-tests").expect("open");
    // Seed one drawer so synthesize has a candidate pool.
    let _id = capture(&registry, "carbon compounds synthesis test");
    let frozen = make_dispatcher(registry, EstatePosture::Frozen);
    let before = (
        std::fs::read(&path).expect("estate file"),
        std::fs::read(format!("{path}-wal")).unwrap_or_default(),
    );

    let result = tools_call(&frozen, "moot_synthesize",
        serde_json::json!({"query": "carbon compounds", "limit": 5}));
    assert!(!is_error(&result),
        "moot_synthesize is a read and must work when frozen; got: {result:?}");
    let after = (
        std::fs::read(&path).expect("estate file"),
        std::fs::read(format!("{path}-wal")).unwrap_or_default(),
    );
    assert_eq!(after, before, "moot_synthesize must leave the estate byte-identical on disk");
    // The frozen refused-set must no longer name moot_synthesize.
    assert!(!is_frozen_refused("moot_synthesize"),
        "moot_synthesize must not be in the refused set after FRZ-3");
    assert!(FROZEN_READ_TOOLS.contains(&"moot_synthesize"),
        "moot_synthesize must be in the explicit read set after FRZ-3");

    let _ = std::fs::remove_file(&path);
}
